content/library/advanced-features/custom-classes.md

```diff
@@ -0,0 +1,150 @@
+---
+title: Custom Classes
+slug: /library/advanced-features/custom-classes
+---
+
+# Custom classes
+
+Developers building larger or more complex applications, or who are porting existing application code over to streamlit may, eventually, want to utilize custom-defined python classes within their script. Common reasons for this include:
+
+1. Defining a custom `@dataclass` to store data within your app
+2. Defining an `Enum` class to represent a fixed set of options.
+3. Defining custom interfaces to external services or databases not covered by Streamlit's `experimental_connection` [API](/library/api-reference/connections/st.experimental_connection).
+
+Because streamlit re-runs your main script from top to bottom after every user interaction, care must be taken when defining custom classes to ensure that a re-definition of the class type during these re-runs will not cause unwanted effects. By following the guidelines on this page, you will be less likely to encounter common pitfalls.
+
+<Collapse title="Table of contents" expanded={true}>
+
+1. [General guildelines](#general-guidelines)
+2. [Enums](#enums)
+
+</Collapse>
+
+## General Guidelines
+
+**For all classes:**
+
+- If possible, move class definitions into their own module file.
+
+  For example:
+
+  ```python
+  # In Main.py ---------------------------------------
+  import streamlit as st
+
+  # MyClass gets re-defined every time Main.py re-runs
+  class MyClass:
+    def __init__(self, var1, var2):
+        ...
+
+  st.write(MyClass("foo", "bar"))
+  ```
+
+  becomes:
+
+  ```python
+  # In my_class.py ---------------------------------------
+  class MyClass:
+    def __init__(self, var1, var2):
+        ...
+
+  # In Main.py -------------------------------------------
+  import streamlit as st
+  from my_class import MyClass
+
+  # MyClass doesn't change identity when Main.py re-runs
+
+  st.write(MyClass("foo", "bar"))
+  ```
+
+  Streamlit only reloads code in imported modules when it detects the code has changed. Thus, each time your page script re-runs it is using the exact same `MyClass` as on the previous execution.
+
+**For classes that store data (like [dataclasses](https://docs.python.org/3/library/dataclasses.html)):**
+
+- Consider defining a custom `__eq__` method based on the data _in_ the class.
+
+  ```python
+  import streamlit as st
+  from dataclasses import dataclass
+
+  @dataclass
+  class MyDataclass:
+    var1: int
+    var2: float
+
+    def __eq__(self, other):
+      # Two instances of MyDataclass are equal if both of their
+      # fields match.
+      return (self.var1, self.var2) == (other.var1, other.var2)
+
+  st.write(MyDataclass(1, 5.5) == MyDataclass(1, 5.5))
+  ```
+
+  The default python `__eq__` implementations for both regular
+  classes as well as `@dataclasses` depend on the in-memory ID of the class or class instance. This can cause comparison problems where two
```

Comment:
This is true for regular classes, but not for dataclasses. It is certainly possible that you would want to ignore certain fields, and so the default `__eq__` might not always be appropriate, but much of the time it works great, because it does exactly this.

> eq: If true (the default), an [__eq__()](https://docs.python.org/3/reference/datamodel.html#object.__eq__) method will be generated. This method compares the class as if it were a tuple of its fields, in order. Both instances in the comparison must be of the identical type.
(From https://docs.python.org/3/library/dataclasses.html#module-contents)

```python
In [2]: @dataclass
   ...: class A:
   ...:     val: int
   ...:     other_val: int
   ...:

In [3]: a = A(1,2)

In [4]: b = A(1,2)

In [5]: c = A(1,3)

In [6]: a == b
Out[6]: True

In [7]: a == c
Out[7]: False
```

Reply:
@blackary Maybe I'm missing a detail, but I think to illustrate the situation here and why all this is a thing, try this:
```python
from dataclasses import dataclass

@dataclass
class A:
    val: int
    other_val: int
a = A(1,2)

@dataclass
class A:
    val: int
    other_val: int
b = A(1,2)

a == b #False
```
Between script runs, if the dataclass gets redefined then instances created on different reruns won't be equal even if instantiated with the same values.

Reply:
Ah! That makes sense, @sfc-gh-dmatthews. If you follow the recommendation of putting your class in a separate module, won't that eliminate this issue?

I can see that issue in the code below with class B, defined locally, but not with imported class A

```python
import streamlit as st
from dataclasses import dataclass
from modules.example import A

if "items" not in st.session_state:
    st.session_state["items"] = [A(1, 2)]


if st.button("Add another"):
    st.session_state["items"].append(A(1, 2))

st.write(st.session_state["items"])

for i in st.session_state["items"]:
    st.write(i)
    st.write(i == st.session_state["items"][0]) # ALL TRUE


@dataclass
class B:
    val: int
    other_val: int


if "items2" not in st.session_state:
    st.session_state["items2"] = [B(1, 2)]

if st.button("Add another 2"):
    st.session_state["items2"].append(B(1, 2))

st.write(st.session_state["items2"])

for i in st.session_state["items2"]:
    st.write(i)
    st.write(i == st.session_state["items2"][0]) # ALL FALSE AFTER FIRST ONE
```

Reply:
Yes, this whole page is about all the different ways to deal with the issue. The first, most-recommended option is to define classes outside the script and import them. It's not shown in the preview yet, since I put them in as suggestions in my other comments, but I'm recommending using headers "Option 1: ...", "Option 2: ....", "Option 3: ..." etc to emphasize that the guide is describing different alternatives to solving the one problem.